addons/mail/controllers/discuss.py

```diff
@@ -1,18 +1,225 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import base64
-import werkzeug.utils
-import werkzeug.wrappers
+import json
 
 from odoo import http
-from odoo.exceptions import AccessError
 from odoo.http import request
+from odoo.tools import consteq
+from odoo.tools.misc import get_lang
+from odoo.tools.translate import _
+from werkzeug.exceptions import NotFound
 
 
 class DiscussController(http.Controller):
 
-    @http.route('/mail/read_followers', type='json', auth='user')
+    # --------------------------------------------------------------------------
+    # Public Pages
+    # --------------------------------------------------------------------------
+
+    @http.route('/chat/<int:channel_id>/<string:invitation_token>', methods=['GET'], type='http', auth='public')
+    def discuss_channel_invitation(self, channel_id, invitation_token, **kwargs):
+        channel_sudo = request.env['mail.channel'].browse(int(channel_id)).sudo().exists()
+        if not channel_sudo or not channel_sudo.uuid or not consteq(channel_sudo.uuid, invitation_token):
+            raise NotFound()
+        response = request.redirect(f'/discuss/channel/{channel_sudo.id}')
+        if request.env['mail.channel.partner']._get_as_sudo_from_request(request=request, channel_id=int(channel_id)):
+            return response
+        if channel_sudo.channel_type == 'chat':
+            raise NotFound()
+        if request.session.uid:
+            channel_sudo.add_members([request.env.user.partner_id.id])
+            return response
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        if not guest:
+            guest = request.env['mail.guest'].sudo().create({
+                'country_id': request.env['res.country'].sudo().search([('code', '=', request.session.get('geoip', {}).get('country_code'))], limit=1).id,
+                'lang': get_lang(request.env).code,
+                'name': _("Guest"),
+                'timezone': request.env['mail.guest']._get_timezone_from_request(request),
+            })
+            # Guest session cookies, every route in this file will make use of them to authenticate
+            # the guest through `_get_as_sudo_from_request` or `_get_as_sudo_from_request_or_raise`.
+            response.set_cookie('mail.guest_id', str(guest.id), httponly=True)
+            response.set_cookie('mail.guest_access_token', guest.access_token, httponly=True)
```

Comment:
One more thing - proliferation of cookies is bad taste, and with GDPR every single one of them needs to be documented for compliance reasons, with the purpose, etc.

So it would be better to use a structured cookie instead of 2 different ones. It's also nice to save some bytes by not making the cookie name longer than strictly necessary. Remember that this stuff is sent for *every single request* that will be made by the browser. It's not like we need users to be able to read the cookies.

```suggestion
            # dgid = "discord guest id"
            response.set_cookie('dgid', f"{guest.id}|{guest.access_token}", httponly=True)
```

then split it when you need the values...

Reply:
Ok for the suggestions, I'll make the changes to cookie post merge, or directly if we need to fix conflict at some point with all that's pending on mergebot.

Reply:
https://github.com/odoo/odoo/pull/75888